### DISCUSSION TITLE
Cobbler tweaks/tips for large-scale operations?

### DISCUSSION THREAD

**Discussion by AndyGrill** [cobbler/cobbler#2980] [Q&A] [answered]
Hi all, we've been using Cobbler in my organization for a few years now, and as one of our sites continues to grow, so too do the system entries in Cobbler.  We're now approaching ~360K system entries in the biggest site, and operations with Cobbler are slowing substantially.

Does anyone here would have any tips or optimization suggestions in the program for such a large data set?  Perhaps there's some config change that's needed in Cobbler that I've missed to help latency, particularly when adding/removing/editing system entries.

Thank you very much.

**Comment by SchoolGuy** [comment, marked as answer]
So the bottleneck is definitely `cobbler sync`. You do want to have fast storage available on your TFTP directory and `/var/lib/cobbler/collections` as that contains the JSON database. The mongodb serializer would be your best bet for performance I guess but the problem is that this is not very well tested.

Apart from that Cobbler uses a lot of templates obviously. However Cobbler is not heavily multi-threaded which means that a faster clock on a single thread might help here. However I don't have data to support that statement.

I would be highly interested as project admin/maintainer to get more feedback about how you use Cobbler and/or what the pain points are. It seems that your installation may very well be the largest one ever set up.

The main topics for Cobbler at the moment are:

- Building the new Web Interface (slow due to limited resources)
- Fixing the `cobbler sync` performance issues
- Further reducing the number of bugs we can find

**Reply by AndyGrill** [reply]
Thank you so much for your response!  We were also thinking SSDs would be helpful for database access, nice to get additional confirmation as such.

In terms of how my organization uses Cobbler, we make use of profiles/kickstarts with systems attached to the profiles for network booting.  We do **not** make use of cobbler sync, at least we don't manually call it in any of our automation.  We have a python automation script to read all the system json files, and pull out the hostname, mac address, and IP address fields from the json file where applicable, and process that into the system declaration portion of /etc/dhcpd.conf.

Where we're seeing massive slowdowns is in adding, removing, or editing systems in cobbler.  Using "cobbler system report" is quick, and "cobbler system find" seems to be no slower than it used to be in this large site.  Where adding/removing/editing used to take ~5 or ~10 seconds to apply in Cobbler, now it takes somewhere between ~2 and ~4 minutes on average, with no visible changes on our end that would explain the sudden drastic slowness.

Restarting httpd didn't help, restarting cobblerd would help for less than 5 minutes before the slowness would return, that's kinda where we're at right now :/

**Reply by SchoolGuy** [reply]
Okay... So our internal data structures are not the most efficient tbh so it can very well be just the Python code. One thing you could try is turning off the duplicate IP detection (key name in `settings.yaml `allow_duplicate_macs`). This may reduce the number of checks done, however I don't expect this to yield a great difference.

The main suspect I have is indeed the much much increased type checking and much stricter validation of data in Cobbler. The reasoning behind those strict checks are various people how "shaped" Cobbler to their desire and then requiring support because Cobbler didn't handle the data well in the very end.

Regarding what you can do now: Well the faster the CPU runs the faster the calculations are done. But this is not really reasonable. Sadly in your exact instance there is nothing you can do except open a performance bug (of which we already have quite some).

Regarding what you can do mid-term: If you would agree we can schedule a Jitsi Session or get in touch via mail and you provide me with some details (numbers & statistics) about what is slow. Then I can have a look if I can correct that inside Cobbler for the whole community via PRs.

**Reply by AndyGrill** [reply]
Thank you for the suggestions! For now, we've got some work to do on our end to get Cobbler updated when we move things to SLES15SP4, so we'll revisit performance and such when we've gotten that done if needed.

Thanks again for the discussion! Assuming we get things updated and there are still problems, I like the idea of jitsi or mail to discuss details :)

**Comment by marbaa** [comment]
Out of curiosity, you have ~360 000 servers/devices in cobbler?

**Reply by AndyGrill** [reply]
correct, that's our largest cobbler instance, others are thousands or tens of thousands
